src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -0,0 +1,209 @@
+
+            // DownloadFromStream each of the artifact sequentially. 
+            // TODO: Should we download them parallely?
+            foreach (ServerBuildArtifact buildArtifact in buildArtifacts)
+            {
+                if (Match(buildArtifact, artifactDefinition))
+                {
+                    executionContext.Output(StringUtil.Loc("RMPreparingToDownload", buildArtifact.Name));
+                    await this.DownloadArtifactAsync(executionContext, buildArtifact, artifactDefinition, localFolderPath, buildClient, xamlBuildClient, buildDefinitionType, buildId);
+                }
+                else
+                {
+                    executionContext.Warning(StringUtil.Loc("RMArtifactMatchNotFound", buildArtifact.Name));
+                }
+            }
+        }
+
+        public IArtifactDetails GetArtifactDetails(IExecutionContext context, AgentArtifactDefinition agentArtifactDefinition)
+        {
+            Trace.Entering();
+
+            ServiceEndpoint vssEndpoint = context.Endpoints.FirstOrDefault(e => string.Equals(e.Name, ServiceEndpoints.SystemVssConnection, StringComparison.OrdinalIgnoreCase));
+            ArgUtil.NotNull(vssEndpoint, nameof(vssEndpoint));
+            ArgUtil.NotNull(vssEndpoint.Url, nameof(vssEndpoint.Url));
+
+            var artifactDetails = JsonConvert.DeserializeObject<Dictionary<string, string>>(agentArtifactDefinition.Details);
+            VssCredentials vssCredentials = ApiUtil.GetVssCredential(vssEndpoint);
+            var tfsUrl = context.Variables.Get(WellKnownDistributedTaskVariables.TFCollectionUrl);
+
+            Guid projectId = context.Variables.System_TeamProjectId ?? Guid.Empty;
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            string relativePath;
+            string accessToken;
+            vssEndpoint.Authorization.Parameters.TryGetValue(EndpointAuthorizationParameters.AccessToken, out accessToken);
+
+            if (artifactDetails.TryGetValue("RelativePath", out relativePath))
+            {
+                return new BuildArtifactDetails
+                {
+                    Credentials = vssCredentials,
+                    RelativePath = artifactDetails["RelativePath"],
+                    Project = projectId.ToString(),
+                    TfsUrl = new Uri(tfsUrl),
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactDetailsIncomplete"));
+            }
+        }
+
+        private bool Match(ServerBuildArtifact buildArtifact, ArtifactDefinition artifactDefinition)
+        {
+            // If this is older , then dont force the name checks. 
+            if (string.Equals(artifactDefinition.Name, AllArtifacts, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            if (string.Equals(artifactDefinition.Name, buildArtifact.Name, StringComparison.OrdinalIgnoreCase))
```

Comment:
are we still supporting editing artifactDefinitionName vai REST?
If not we should remove all this logic in the new agent